Source/CesiumRuntime/Private/CesiumGeoreferenceComponent.cpp

```diff
@@ -253,223 +476,164 @@ void UCesiumGeoreferenceComponent::HandleGeoreferenceUpdated() {
-            CesiumGeospatial::Ellipsoid::WGS84));
-    glm::dmat4 endEnuToEcef =
-        CesiumGeospatial::Transforms::eastNorthUpToFixedFrame(
-            targetEcef,
-            CesiumGeospatial::Ellipsoid::WGS84);
-    // glm::dmat4 startEcefToEnu = glm::affineInverse(
-    //    this->Georeference->ComputeEastNorthUpToEcef(this->_actorToECEF[3]));
-    // glm::dmat4 endEnuToEcef =
-    //    this->Georeference->ComputeEastNorthUpToEcef(targetEcef);
-
-    this->_actorToECEF = endEnuToEcef * startEcefToEnu * this->_actorToECEF;
+  logVector("_setECEF _currentEcef ", _currentEcef);
+  logVector("_setECEF   targetEcef ", targetEcef);
+  _debugLogState();
+
+  if (!IsValid(this->Georeference)) {
+    UE_LOG(
+        LogCesium,
+        Warning,
+        TEXT(
+            "CesiumGeoreferenceComponent %s does not have a valid Georeference"),
+        *this->GetName());
+    return;
   }
 
-  this->_updateActorToUnrealRelativeWorldTransform();
-  this->_setTransform(this->_actorToUnrealRelativeWorld);
+  const glm::dmat3 oldActorRotation = _getRotationFromActor();
+  glm::dmat3 newActorRotation = oldActorRotation;
+  const glm::dvec3 newRelativeLocation = _computeRelativeLocation(targetEcef);
 
-  // In this case the ground truth is the newly updated _actorToECEF
-  // transformation, so it will be more accurate to compute the new Unreal
-  // locations this way (as opposed to _updateAbsoluteLocation /
-  // _updateRelativeLocation).
-  this->_relativeLocation = this->_actorToUnrealRelativeWorld[3];
-  this->_absoluteLocation =
-      this->_relativeLocation + this->_worldOriginLocation;
+  if (!maintainRelativeOrientation) {
 
-  // If the transform needs to be snapped to the tangent plane, do it here.
-  if (this->_autoSnapToEastSouthUp) {
-    this->SnapToEastSouthUp();
-  }
+    // When NOT maintaining relative orientation, just
+    // update the ECEF position with the new values
+
+    this->ECEF_X = targetEcef.x;
+    this->ECEF_Y = targetEcef.y;
+    this->ECEF_Z = targetEcef.z;
+    this->_currentEcef = targetEcef;
+
+  } else {
+
+    // When maintaining the relative orientation, compute
+    // - the Unreal-to-ENU transform at the old position
+    // - the ENU-to-Unreal transform at the new position
+    // - use this to compute the new actor rotation
+    // - update the ECEF position with the new values
+
+    // Note: this probably degenerates when starting at or moving to either of
+    // the poles
 
-  // Update component properties
-  this->_updateDisplayECEF();
+    const glm::dvec3 currentRelativeLocation =
+        _computeRelativeLocation(_currentEcef);
+    const glm::dmat3 currentUnrealToEnu =
+        glm::inverse(this->Georeference->ComputeEastNorthUpToUnreal(
+            currentRelativeLocation));
+    const glm::dmat3 newEnuToUnreal(
+        this->Georeference->ComputeEastNorthUpToUnreal(newRelativeLocation));
+    newActorRotation = newEnuToUnreal * currentUnrealToEnu * oldActorRotation;
```

Comment:
Would it be more consistent to compute a glm::rotation here? Similar to how it's done in line 107

Reply:
Intuitively, it sounds reasonable to compute the ellipsoid normals at these points, and use the rotation between these (as computed with `glm::rotation`) for the update. I'm not yet 100% sure whether this will really retain the "full" orientation (meaning that I'm not sure whether it prevents the object from appearing to be rotated around the ellipsoid normal direction), but ... after all, I think this should work, becasue this should be exactly "the orientation change" between both points. I'll try to think this through more thoroughly (or just try it out...).

Reply:
Thanks, done in https://github.com/CesiumGS/cesium-unreal/pull/518/commits/d5845b4f94ee4a27e83c54c8b8c244dc320b0066 . It indeed looks much cleaner and simpler (and seems to work well, although I "only" tried it out, though, and didn't do the maths...)

Unfortunately, I'm a bit held up with some other tasks that came in unexpectedly, but will try to focus on the crashes next. I'm regularly testing this with a level that I created in the `main` branch, though, and didn't see crashes (that hadn't been fixed in the meantime) until now. If you have any steps to reproduce them, that would be appreciated, otherwise, I'll try to "provoke" them in some way, maybe based on what I can derive from the traces...